### PR TITLE
Create notes app. Adding notes, deleting notes and updating notes work.

### PR DESCRIPTION
**Pull request by tamokakhidze**

### CODE REVIEW

File: NotesApp/CoreData/CoreDataManager.swift [hunk 1]
```diff
@@ -6,3 +6,76 @@
 //
 
 import Foundation
```
Comment (GujMeister) [inline NotesApp/CoreData/CoreDataManager.swift:8]: Remove "import Foundation"

File: NotesApp/CoreData/CoreDataManager.swift [hunk 2]
```diff
@@ -6,3 +6,76 @@
...
+protocol CoreDataManagerDelegate: AnyObject {
+    func reloadTableView()
+}
+
+class CoreDataManager {
+    
+    public static let shared = CoreDataManager()
+    var viewModel = ViewModel()
+    var notes: [NSManagedObject] = []
+    private init() {}
```
Comment (GujMeister) [inline NotesApp/CoreData/CoreDataManager.swift:21]: Separate initialisation method from other properties

File: NotesApp/Customs/BlackAlertController.swift [hunk 3]
```diff
@@ -7,14 +7,113 @@
 
 import UIKit
 
+//MARK: - BlackAlertController
 class BlackAlertController: UIView {
```
Comment (GujMeister) [inline NotesApp/Customs/BlackAlertController.swift:11]: Try to make classes that will not be subclassed "final"

File: NotesApp/Scenes/MainVC/ViewController.swift [hunk 4]
```diff
@@ -6,24 +6,174 @@
...
+    func deleteNote(_ note: NSManagedObject, at indexPath: IndexPath) {
+        viewModel.deleteNote(note, at: indexPath)
+        tableView.deleteSections(IndexSet(integer: indexPath.section), with: .middle)
+    }
+    
+}
 
+//MARK: - UITableViewDelegate
+extension ViewController: UITableViewDelegate {
+
```
Comment (GujMeister) [inline NotesApp/Scenes/MainVC/ViewController.swift:152]: Remove whitespace on 152nd line

File: NotesApp/Scenes/MainVC/ViewModel.swift [hunk 5]
```diff
@@ -6,3 +6,19 @@
 //
 
 import Foundation
+import CoreData
+
+//MARK: - ViewModel
+class ViewModel {
+
```
Comment (GujMeister) [inline NotesApp/Scenes/MainVC/ViewModel.swift:13]: Remove whitespace on 13th line

**Review by GujMeister — CHANGES_REQUESTED** [review]
Understandable code with minor stylistic mistakes which should be fixed before resubmitting

**Review by GujMeister — APPROVED** [review]
Thanks for resolving the problems stated above.

**Review by AnaIoramashvili — CHANGES_REQUESTED** [review]
In NoteCell please change the font size. The fond size should be: 16

**Review by AnaIoramashvili — APPROVED** [review]
Great Job!

File: NotesApp/Customs/BlackAlertController.swift [hunk 6]
```diff
@@ -7,14 +7,113 @@
...
+//MARK: - BlackAlertController
+final class BlackAlertController: UIView {
 
-    /*
-    // Only override draw() if you perform custom drawing.
-    // An empty implementation adversely affects performance during animation.
-    override func draw(_ rect: CGRect) {
-        // Drawing code
+    //MARK: - Properties
+    private lazy var icon: UIImageView = {
```
Comment (AniDzamelashvili) [inline NotesApp/Customs/BlackAlertController.swift:14]: Do you really need to have lazy properties for icon and alertmessahe?

File: NotesApp/Customs/BlackAlertController.swift [hunk 7]
```diff
@@ -7,14 +7,113 @@
...
+        buttonsStackView.addArrangedSubview(cancelButton)
+        buttonsStackView.addArrangedSubview(deleteButton)
+        addSubview(buttonsStackView)
+        
+        NSLayoutConstraint.activate([
+            icon.topAnchor.constraint(equalTo: topAnchor, constant: 20),
+            icon.centerXAnchor.constraint(equalTo: centerXAnchor),
+            icon.heightAnchor.constraint(equalToConstant: 50),
+            icon.widthAnchor.constraint(equalToConstant: 50),
+
```
Comment (AniDzamelashvili) [inline NotesApp/Customs/BlackAlertController.swift:94]: Please have seperate functions as I mentioned above

File: NotesApp/Customs/NoteCell.swift
Comment (AniDzamelashvili) [inline NotesApp/Customs/NoteCell.swift:1]: Same comment here, not write everything in init.

File: NotesApp/Customs/NoteCell.swift [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        titleLabel.heightAnchor.constraint(equalToConstant: 67).isActive = true
+        titleLabel.centerYAnchor.constraint(equalTo: centerYAnchor).isActive = true
+        titleLabel.widthAnchor.constraint(lessThanOrEqualToConstant: 290).isActive = true
+        titleLabel.numberOfLines = 2
+        titleLabel.font = .systemFont(ofSize: 24, weight: .medium)
+        return titleLabel
+    }
+    
+    private func configureDeleteButton() -> UIButton {
+        let deleteButton = UIButton()
```
Comment (AniDzamelashvili) [inline NotesApp/Customs/NoteCell.swift:46]: why are you preferring having properties like this? you were writing inBlackALertView properties with closure. currently inside this you are having everything like adding subviews, constraints ...

File: NotesApp/Scenes/AddNoteVC/AddNoteViewController.swift [hunk 9]
```diff
@@ -6,24 +6,149 @@
...
+        if let description = descriptionText {
+            descriptionTextField.text = description
+        }
+        
+        configureTitleTextField()
+        configureDescriptionTextField()
+        configureSaveButton()
+    }
+    
+    private func configureTitleTextField() {
```
Comment (AniDzamelashvili) [inline NotesApp/Scenes/AddNoteVC/AddNoteViewController.swift:66]: You were writing inBlackALertView properties with closure. currently inside this you are having everything like adding subviews, constraints ... take a look to inBlackALertView file

File: NotesApp/Scenes/AddNoteVC/AddNoteViewController.swift [hunk 10]
```diff
@@ -6,24 +6,149 @@
...
-        // Get the new view controller using segue.destination.
-        // Pass the selected object to the new view controller.
+extension AddNoteViewController: UITextFieldDelegate {
+    func textFieldDidEndEditing(_ textField: UITextField) {
+        guard let managedContext = (UIApplication.shared.delegate as? AppDelegate)?.persistentContainer.viewContext,
+              let noteID = noteID,
+              let note = managedContext.object(with: noteID) as? NSManagedObject else {
+            return
+        }
+
```
Comment (AniDzamelashvili) [inline NotesApp/Scenes/AddNoteVC/AddNoteViewController.swift:139]: fix indentation

File: NotesApp/Scenes/MainVC/ViewController.swift [hunk 11]
```diff
@@ -6,24 +6,184 @@
...
+}
 
-    // In a storyboard-based application, you will often want to do a little preparation before navigation
-    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
-        // Get the new view controller using segue.destination.
-        // Pass the selected object to the new view controller.
+//MARK: - UITableViewDataSource
+extension ViewController: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        1
```
Comment (AniDzamelashvili) [inline NotesApp/Scenes/MainVC/ViewController.swift:105]: avoid using hardcoded numbers

**Pull request merged**